### ISSUE TITLE
WinRM file upload/download should be separate gem

### ISSUE THREAD

**Issue by sneal** [WinRb/WinRM#114]
@zenchild @mwrock @maxlinc Thoughts about splitting file operations (upload) out into a separate gem?

Pros:
- SRP.
- It would allow us to release WinRM gem 1.3 sooner.
- Easier to remove later if implement a better solution.

Cons:
- Yet another gem/dependency
- More work

**Comment by mwrock** [comment]
I've been thinking about this too. I think its a good idea mainly for keeping tooling on top of the protocol separate from the actual protocol implementation itself.

I'm happy to own this or assist in the decoupling. Should have cycles after the weekend.

**Comment by maxlinc** [comment]
I don't have a strong opinion at this point. If it means 1.3 sooner that's good, and since Vagrant currently has its own copy of the file upload/download logic it won't block that - can just create an issue in Vagrant to remove the custom file manager and replace with the winrm-files gem when it's stable.

In the long run it might be good if you had a generic FileManager interface with pluggable protocols, so that you can provide a gem with a basic but inefficient protocol (i.e. `winrm-files-base64`), but others could create alternative solutions, like the `winrm-files-rsync`, `winrm-files-bits` or `winrm-files-git` (all ideas that came up in Vagrant/Packer discussions).

E.g.:
```ruby
require 'winrm'
require 'winrm/files/rsync'

endpoint = 'http://mywinrmhost:5985/wsman'
krb5_realm = 'EXAMPLE.COM'
winrm = WinRM.new(endpoint, :kerberos, :realm => krb5_realm)
default_fm = winrm.file_manager
rsync_fm = winrm.file_manager(:rsync)

require 'benchmark'
# Note that with bmbm the "reheasal" time will be
# for a new upload, and then the benchmark time
# will be for a re-upload.
Benchmark.bmbm(20) do |x|
  x.report("Default (base64):")   { default_fm.upload('./folder') }
  x.report("Rsync:") { rsync_fm.upload('./folder') }
end
```

**Comment by zenchild** [comment]
I like the idea of keeping the protocol as skinny as possible and having a winrm-extensions gem or something like that. It makes it simpler to digest for the brain when you're trying to iterate on either. Your dealing with an interface instead of an entire code-base.

**Comment by mwrock** [comment]
+1 to the above comments, but I do think it would be best to not have the separate repo be a generic extensions repo. I'd be concerned that could possibly muddy the waters. It would be better to keep the repos targeted to the problem being solved. I have 2 projects on my backlog that could fit in the extensions category but would be best left separate: running winrm commands via a scheduled task and managing reboots to the remote endpoint keeping a "percieved" persistent connection.

**Comment by sneal** [comment]
@zenchild Can you create us a new `winrm-fs` repository under the winrb organization and make both @mwrock and @sneal repo owners?

@maxlinc The generic FileManager interface is an interesting idea. I think that's something we can add later if needed.

**Comment by zenchild** [comment]
winrm-fs is setup

**Comment by sneal** [comment]
Fixed by #116
